# Data Preprocessing - Feild Asset Health Monitor Project

%autoreload 2 is the magic that makes this workflow pleasant: you develop functions in VSC in the .py file, and the notebook always runs the latest version.

In [2]:
# re-imports your src modules on every cell run — edit the .py, rerun the cell, no kernel restart
%load_ext autoreload
%autoreload 2          

#### raw loading the data

In [3]:
from fahm import preprocessing as pp

cfg = pp.load_config("../configs/config.yaml")

In [4]:
print(cfg["paths"]["raw_csv"])

/home/leoadmin/Documents/GitHub/FAHM_Project/data/raw/MetroPT3(AirCompressor).csv


In [5]:
df = pp.load_raw(cfg)              # work happens in src
df.head()                          # notebook does the LOOKING

,timestamp,TP2,TP3,H1,DV_pressure,Reservoirs,Oil_temperature,Motor_current,COMP,DV_eletric,Towers,MPG,LPS,Pressure_switch,Oil_level,Caudal_impulses
0,2020-02-01 00:00:00,-0.012,9.358,9.340,-0.024,9.358,53.600,0.0400,1,0,1,1,0,1,1,1
1,2020-02-01 00:00:10,-0.014,9.348,9.332,-0.022,9.348,53.675,0.0400,1,0,1,1,0,1,1,1
2,2020-02-01 00:00:19,-0.012,9.338,9.322,-0.022,9.338,53.600,0.0425,1,0,1,1,0,1,1,1
3,2020-02-01 00:00:29,-0.012,9.328,9.312,-0.022,9.328,53.425,0.0400,1,0,1,1,0,1,1,1
4,2020-02-01 00:00:39,-0.012,9.318,9.302,-0.022,9.318,53.475,0.0400,1,0,1,1,0,1,1,1


In [6]:
df.tail()

,timestamp,TP2,TP3,H1,DV_pressure,Reservoirs,Oil_temperature,Motor_current,COMP,DV_eletric,Towers,MPG,LPS,Pressure_switch,Oil_level,Caudal_impulses
1516943,2020-09-01 03:59:10,-0.014,8.918,8.906,-0.022,8.918,59.675,0.0425,1,0,1,1,0,1,1,1
1516944,2020-09-01 03:59:20,-0.014,8.904,8.888,-0.020,8.904,59.600,0.0450,1,0,1,1,0,1,1,1
1516945,2020-09-01 03:59:30,-0.014,8.890,8.876,-0.022,8.892,59.600,0.0425,1,0,1,1,0,1,1,1
1516946,2020-09-01 03:59:40,-0.012,8.876,8.864,-0.022,8.878,59.550,0.0450,1,0,1,1,0,1,1,1
1516947,2020-09-01 03:59:50,-0.014,8.860,8.848,-0.022,8.864,59.475,0.0425,1,0,1,1,0,1,1,1


In [7]:
df.info(memory_usage="deep")

<class 'pandas.DataFrame'>
RangeIndex: 1516948 entries, 0 to 1516947
Data columns (total 16 columns):
 #   Column           Non-Null Count    Dtype         
---  ------           --------------    -----         
 0   timestamp        1516948 non-null  datetime64[us]
 1   TP2              1516948 non-null  float64       
 2   TP3              1516948 non-null  float64       
 3   H1               1516948 non-null  float64       
 4   DV_pressure      1516948 non-null  float64       
 5   Reservoirs       1516948 non-null  float64       
 6   Oil_temperature  1516948 non-null  float64       
 7   Motor_current    1516948 non-null  float64       
 8   COMP             1516948 non-null  int8          
 9   DV_eletric       1516948 non-null  int8          
 10  Towers           1516948 non-null  int8          
 11  MPG              1516948 non-null  int8          
 12  LPS              1516948 non-null  int8          
 13  Pressure_switch  1516948 non-null  int8          
 14  Oil_level    

#### how many missing values?

In [8]:
df.isna().sum()

timestamp          0
TP2                0
TP3                0
H1                 0
DV_pressure        0
Reservoirs         0
Oil_temperature    0
Motor_current      0
COMP               0
DV_eletric         0
Towers             0
MPG                0
LPS                0
Pressure_switch    0
Oil_level          0
Caudal_impulses    0
dtype: int64

In [9]:
df["timestamp"].diff().value_counts()

timestamp
0 days 00:00:10    1337521
0 days 00:00:09     128277
0 days 00:00:12      38321
0 days 00:00:13       7988
0 days 00:00:11       4471
                    ...   
0 days 03:45:10          1
0 days 01:19:09          1
0 days 00:05:41          1
0 days 04:22:23          1
0 days 01:55:04          1
Name: count, Length: 337, dtype: int64

In [10]:
# your point 5, now unlocked: interval distribution (needs datetime — which you now have)
df[TIMESTAMP := "timestamp"].diff().value_counts().head(10)

timestamp
0 days 00:00:10    1337521
0 days 00:00:09     128277
0 days 00:00:12      38321
0 days 00:00:13       7988
0 days 00:00:11       4471
0 days 00:00:21         10
0 days 00:00:19          5
0 days 00:00:22          4
0 days 00:00:20          3
0 days 00:00:17          3
Name: count, dtype: int64

#### Lets a have a look on autoviz eda, knowing that a manual eda is a must for this kind of data

In [11]:
# importing plotting functions:
from fahm import plotting as pl

In [12]:
%matplotlib inline
pl.auto_eda_pdf(df, "../docs/autoviz_report.pdf")

Imported v0.1.905. Please call AutoViz in this sequence:
    AV = AutoViz_Class()
    %matplotlib inline
    dfte = AV.AutoViz(filename, sep=',', depVar='', dfte=None, header=0, verbose=1, lowess=False,
               chart_format='svg',max_rows_analyzed=150000,max_cols_analyzed=30, save_plot_dir=None)
    Since nrows is smaller than dataset, loading random sample of 150000 rows into pandas...
Shape of your Data Set loaded: (150000, 15)
#######################################################################################
######################## C L A S S I F Y I N G  V A R I A B L E S  ####################
#######################################################################################
Classifying variables in data set...
    Number of Numeric Columns =  7
    Number of Integer-Categorical Columns =  0
    Number of String-Categorical Columns =  0
    Number of Factor-Categorical Columns =  0
    Number of String-Boolean Columns =  0
    Number of Numeric-Boolean Columns =  8

,Data Type,Missing Values%,Unique Values%,Minimum Value,Maximum Value,DQ Issue
TP2,float64,0.000000,NA,-0.032000,10.624000,Column has 25063 outliers greater than upper bound (-0.00) or lower than lower bound(-0.02). Cap them or remove them.
TP3,float64,0.000000,NA,0.878000,10.288000,Column has 461 outliers greater than upper bound (10.99) or lower than lower bound(7.02). Cap them or remove them.
H1,float64,0.000000,NA,-0.034000,10.288000,"Column has 22290 outliers greater than upper bound (11.03) or lower than lower bound(6.63). Cap them or remove them., Column has a high correlation with ['TP2']. Consider dropping one of them."
DV_pressure,float64,0.000000,NA,-0.032000,7.448000,Column has 11162 outliers greater than upper bound (-0.01) or lower than lower bound(-0.03). Cap them or remove them.
Reservoirs,float64,0.000000,NA,0.862000,10.290000,"Column has 462 outliers greater than upper bound (10.98) or lower than lower bound(7.02). Cap them or remove them., Column has a high correlation with ['TP3']. Consider dropping one of them."
Oil_temperature,float64,0.000000,NA,17.375000,89.050000,Column has 414 outliers greater than upper bound (81.97) or lower than lower bound(43.18). Cap them or remove them.
Motor_current,float64,0.000000,NA,0.020000,8.680000,No issue
COMP,int8,0.000000,0,0.000000,1.000000,No issue
DV_eletric,int8,0.000000,0,0.000000,1.000000,No issue
Towers,int8,0.000000,0,0.000000,1.000000,No issue


Number of All Scatter Plots = 28
All Plots done
Time to run AutoViz = 14 seconds 

 ###################### AUTO VISUALIZATION Completed ########################


PosixPath('../docs/autoviz_report.pdf')

In [13]:
from IPython.display import HTML

HTML('<a href="../docs/autoviz_report.pdf" target="_blank">📄 Click here to view the full report (PDF)</a>')

#### making manual EDA

In [14]:
fig_ad = pl.plot_analog_distributions(df)

In [15]:
fig_adl = pl.plot_analog_distributions(df, log=True)

In [16]:
summary = pl.digital_summary(df)
summary

,n_zeros,n_ones,n_other,frac_active
signal,,,,
COMP,247328,1269620,0,0.8370
DV_eletric,1273310,243638,0,0.1606
Towers,121586,1395362,0,0.9198
MPG,253840,1263108,0,0.8327
LPS,1511760,5188,0,0.0034
Pressure_switch,12990,1503958,0,0.9914
Oil_level,145391,1371557,0,0.9042
Caudal_impulses,95406,1421542,0,0.9371


In [17]:
fig_ds = pl.plot_digital_summary(summary)

#### saving plots

In [18]:
pl.save_fig(fig_ad, "analog_distributions", cfg)
pl.save_fig(fig_adl, "analog_distributions_log", cfg)
pl.save_fig(fig_ds, "digital_frac_active", cfg)

PosixPath('/home/leoadmin/Documents/GitHub/FAHM_Project/docs/figures/digital_frac_active.png')

#### Some investigations to have a better insight

## EDA Findings: Questions & Answers

---

### Q1 — Which analog sensors are multi-modal, and why (physically)?

**TP2, H1, Motor_current, and DV_pressure.**

**Rule of thumb: N modes ⇔ N relevant machine states for that sensor!**

Which means that each sensor has as many peaks as it has distinct machine states,
the modes map to the machine's operating stated:

- **TP2 and H1** — 2 modes: ~0 bar at idle vs. 8–10 bar under load.
- **Motor_current** — 3 modes, **~0 A = motor off.**
  **~4 A = offloaded running** (motor spinning, intake closed, no compression — TP2 = 0, COMP = 1)
  as we know from the paper, there is a cross-referencing between TP2 and COMP in time.
  **~5.5–6.2 A = under load** (compressing — TP2 sawtooth 8–10 bar, COMP = 0).
  
  note: the large 3.9 A histogram mode (on 6.4 after 16:15) is large because OFFLOADED running lasts
  minutes and not loading, which can be confusing.*
- **DV_pressure** — 2 modes: ~0 while the compressor works under load vs.
  1–2 bar during tower-discharge events.



In [19]:
import pandas as pd

In [20]:
# 1. locate the extreme value
idx = df["DV_pressure"].idxmax()
df.loc[idx - 3 : idx + 3]          # the row + 3 neighbors each side (10s apart)

,timestamp,TP2,TP3,H1,DV_pressure,Reservoirs,Oil_temperature,Motor_current,COMP,DV_eletric,Towers,MPG,LPS,Pressure_switch,Oil_level,Caudal_impulses
481274,2020-04-06 14:18:00,9.820,9.678,-0.014,1.244,9.678,71.450,5.9625,0,1,0,0,0,1,1,1
481275,2020-04-06 14:18:09,9.824,9.682,-0.012,4.328,9.680,71.275,6.0300,0,1,0,0,0,1,1,1
481276,2020-04-06 14:18:19,8.558,9.696,-0.014,3.616,9.694,71.350,5.6800,0,1,1,0,0,1,1,1
481277,2020-04-06 14:18:29,10.100,9.798,-0.014,9.844,9.796,71.375,6.0700,0,1,1,0,0,1,1,1
481278,2020-04-06 14:18:39,10.258,9.970,-0.014,3.330,9.966,71.525,6.0725,0,1,1,0,0,1,1,1
481279,2020-04-06 14:18:49,10.406,10.122,-0.014,2.316,10.120,71.450,6.2775,0,1,1,0,0,1,1,1
481280,2020-04-06 14:18:59,9.672,10.172,-0.012,0.004,10.168,71.425,6.1000,0,1,1,0,0,1,1,1


In [21]:
# 2. see its neighborhood in time
t = df.loc[idx, "timestamp"]
fig_s_tl = pl.plot_sensor_timeline(df, "DV_pressure",t - pd.Timedelta("2h"), t + pd.Timedelta("2h"))

In [22]:
pl.save_fig(fig_s_tl, "DV_pressure_isolate_peak", cfg)

PosixPath('/home/leoadmin/Documents/GitHub/FAHM_Project/docs/figures/DV_pressure_isolate_peak.png')

In [23]:
for s in ["TP2", "Motor_current", "COMP"]: fig = pl.plot_sensor_timeline(df, s, "2020-04-06 14:00", "2020-04-06 15:00")

**Investigated via multi-sensor timeline (DV_pressure, TP2, Motor_current, COMP):** 
real operational episode, not a glitch.

Ten minutes of abnormally rapid loaded cycling (~1 cycle/min, motor ~6A, COMP=0) beginning right after a ~1h50m recording gap,
terminating in the 9.8 bar discharge spike at 14:18, followed by offloaded running (TP2=0, COMP=1, motor ~3.9A) with brief load bursts.

Consistent with a maintenance/test session. Side finding: 
corrected the Motor_current mode mapping — ~3.9A = offloaded running (not sustained load), ~6A = under load — matching the paper's documented levels.

### Q2 — Which sensor pair has near-identical distributions?

**There is a similar distribution between the sensors TP3 and Reservoirs.**  They measure the same connected air volume at two
different points, as we know from the paper notes - visible in the correlation heatmap and in the raw values (identical to 3
decimals in head()).

→ consistency check mean(|TP3 − Reservoirs|) < ε.

In [25]:
display((df["TP3"] - df["Reservoirs"]).abs().describe())

count    1.516948e+06
mean     1.862102e-03
std      1.712118e-03
min      0.000000e+00
25%      0.000000e+00
50%      2.000000e-03
75%      2.000000e-03
max      1.820000e-01
dtype: float64

**Quantified: mean |TP3 − Reservoirs| = 0.0019 bar** 
- median 0.002, max 0.182 over 7 months; differences quantized in 0.002 bar steps — the sensors' resolution.
- The max is consistent with fast-pressure transients sampled 10s apart at two circuit points. 
- **conclusion: → ε = 0.02 bar** set in config (≈5× observed mean): trips on real divergence, passes all observed normal operation.

---

### Q3 — Digital signals: purity and surprising activity fractions

**`n_other == 0` for all eight signals on the full 1,516,948 rows** — the
full-data proof that the int8 cast in D04 was safe.

Surprising fractions:

- **Oil_level = 0.904 active**, yet the docs say active = "oil below
  expected level". A machine running with an oil fault for ~90% of seven
  months is implausible → working hypothesis: **polarity inverted in this
  data release** (active = OK; the interesting 10% is the *inactive* time).
  **Test (step 5):** does the signal flip state around the documented June
  oil-leak failure window?
- **Pressure_switch = 0.991 active** — documented as detecting tower
  discharge *events*, which are brief; an event detector active 99% of the
  time suggests the same inverted/level-type behavior.

Bonus finding: **COMP (0.837) + DV_eletric (0.161) ≈ 0.998** — the two
valve signals are almost perfectly antiphase, independently confirming the
documented relationship.

→ **Feeds step 4:** check `mean(COMP + DV_eletric) ≈ 1`.

---

### Q4 — Data errors vs. machine behavior

- **TP2 idle mode ≈ −0.012 bar** — negative pressure is physically
  impossible, so this is sensor **zero-offset (calibration)**, not an
  error. Consequence: analog range checks must tolerate small negative
  pressures, or they will false-alarm on normal idle.

- **DV_pressure: single ~9.8 bar sample in 1.5M rows** — investigated via
  multi-sensor timeline (DV_pressure, TP2, Motor_current, COMP on
  2020-04-06): **real operational episode, not a glitch.**
  The full story: a ~1h50m recording gap ends at 14:09 (revealed by the
  plot window truncating itself); then ten minutes of abnormally rapid
  loaded cycling (~1 cycle/min, motor ~6 A, COMP = 0) — unusually intense,
  high-demand operation; terminating in the 9.8 bar discharge spike at
  ~14:18; followed by offloaded running (TP2 = 0, COMP = 1, motor ~3.9 A)
  with brief load bursts every ~10 min. Consistent with a
  **maintenance/test session** (e.g., brake tests + vent + standby).
  Dated 12 days before failure F1 — probably unrelated to it.
  Classified as rare *behavior*, not data error.

- **Motor_current tail > 8 A** (counts < 10): rare start-up transients —
  plausible behavior.

- **TP3 / Reservoirs thin low tail down to ~0.7 bar**: rare deep
  depressurizations — likely post-failure moments; to be cross-checked
  against the failure windows in step 5.

### A quick 